File: src/rtmx/auth/oidc.py
"""OIDC authentication client for RTMX.

Implements PKCE-based authentication flow for CLI applications
using Zitadel as the identity provider.

Security features:
- PKCE (Proof Key for Code Exchange) to prevent authorization code interception
- Secure token storage in system keychain
- Short-lived access tokens with refresh capability
- No client secret (public client)

Flow:
1. Generate PKCE code verifier and challenge
2. Start local callback server
3. Open browser to Zitadel authorize endpoint
4. Receive authorization code via callback
5. Exchange code + verifier for tokens
6. Store tokens securely
"""

from __future__ import annotations

import base64
import hashlib
import http.server
import json
import secrets
import socketserver
import threading
import urllib.parse
import webbrowser
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from pathlib import Path
from typing import Any

# Optional keyring support
try:
    import keyring

    HAS_KEYRING = True
except ImportError:
    HAS_KEYRING = False


@dataclass
class AuthConfig:
    """OIDC authentication configuration.

    Attributes:
        provider: Identity provider name (e.g., "zitadel")
        issuer: OIDC issuer URL (e.g., "https://auth.rtmx.ai")
        client_id: Public client ID for CLI
        scopes: Requested OAuth scopes
        callback_port: Local port for callback server
    """

    provider: str = "zitadel"
    issuer: str = "https://auth.rtmx.ai"
    client_id: str = "rtmx-cli"
    scopes: list[str] = field(
        default_factory=lambda: ["openid", "profile", "email", "offline_access"]
    )
    callback_port: int = 8765

    @property
    def authorization_endpoint(self) -> str:
        """Get authorization endpoint URL."""
        return f"{self.issuer}/oauth/v2/authorize"

    @property
    def token_endpoint(self) -> str:
        """Get token endpoint URL."""
        return f"{self.issuer}/oauth/v2/token"

    @property
    def userinfo_endpoint(self) -> str:
        """Get userinfo endpoint URL."""
        return f"{self.issuer}/oidc/v1/userinfo"

    @property
    def redirect_uri(self) -> str:
        """Get callback redirect URI."""
        return f"http://localhost:{self.callback_port}/callback"

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> AuthConfig:
        """Create config from dictionary."""
        return cls(
            provider=data.get("provider", "zitadel"),
            issuer=data.get("issuer", "https://auth.rtmx.ai"),
            client_id=data.get("client_id", "rtmx-cli"),
            scopes=data.get("scopes", ["openid", "profile", "email", "offline_access"]),
            callback_port=data.get("callback_port", 8765),
        )


@dataclass
class TokenInfo:
    """OAuth token information.

    Attributes:
        access_token: Current access token
        refresh_token: Token for refreshing access (optional)
        expires_at: Expiration timestamp
        token_type: Token type (usually "Bearer")
        id_token: OIDC ID token for user info (optional)
    """

    access_token: str
    refresh_token: str = ""
    expires_at: datetime = field(default_factory=datetime.now)
    token_type: str = "Bearer"
    id_token: str = ""

    @property
    def is_expired(self) -> bool:
        """Check if access token is expired."""
        return datetime.now() >= self.expires_at

    @property
    def is_refreshable(self) -> bool:
        """Check if tokens can be refreshed."""
        return bool(self.refresh_token)

    def to_dict(self) -> dict[str, Any]:
        """Convert to dictionary for storage."""
        return {
            "access_token": self.access_token,
            "refresh_token": self.refresh_token,
            "expires_at": self.expires_at.isoformat(),
            "token_type": self.token_type,
            "id_token": self.id_token,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> TokenInfo:
        """Create from dictionary."""
        expires_at = data.get("expires_at", "")
        if isinstance(expires_at, str):
            expires_at = datetime.fromisoformat(expires_at)
        return cls(
            access_token=data.get("access_token", ""),
            refresh_token=data.get("refresh_token", ""),
            expires_at=expires_at,
            token_type=data.get("token_type", "Bearer"),
            id_token=data.get("id_token", ""),
        )


# PKCE utilities


def generate_code_verifier() -> str:
    """Generate a cryptographically random code verifier for PKCE.

    Returns:
        43-128 character URL-safe string
    """
    return secrets.token_urlsafe(64)[:128]


def generate_code_challenge(verifier: str) -> str:
    """Generate code challenge from verifier using S256 method.

    Args:
        verifier: PKCE code verifier

    Returns:
        Base64url-encoded SHA256 hash
    """
    digest = hashlib.sha256(verifier.encode("ascii")).digest()
    return base64.urlsafe_b64encode(digest).rstrip(b"=").decode("ascii")


# Token storage


def _get_token_path() -> Path:
    """Get path for file-based token storage (fallback)."""
    config_dir = Path.home() / ".rtmx"
    config_dir.mkdir(parents=True, exist_ok=True)
    return config_dir / "tokens.json"


def _store_tokens(tokens: TokenInfo, config: AuthConfig) -> None:
    """Store tokens securely.

    Uses system keychain if available, falls back to encrypted file.
    """
    if HAS_KEYRING:
        keyring.set_password(
            "rtmx",
            f"{config.provider}:{config.client_id}",
            json.dumps(tokens.to_dict()),
        )
    else:
        # Fallback to file storage (less secure)
        token_path = _get_token_path()
        token_path.write_text(json.dumps(tokens.to_dict()))
        token_path.chmod(0o600)  # Restrict permissions


def _load_tokens(config: AuthConfig) -> TokenInfo | None:
    """Load stored tokens.

    Returns:
        TokenInfo if found and valid, None otherwise
    """
    if HAS_KEYRING:
        try:
            stored = keyring.get_password("rtmx", f"{config.provider}:{config.client_id}")
            if stored:
                return TokenInfo.from_dict(json.loads(stored))
        except Exception:
            pass
    else:
        token_path = _get_token_path()
        if token_path.exists():
            try:
                return TokenInfo.from_dict(json.loads(token_path.read_text()))
            except Exception:
                pass
    return None


def _clear_tokens(config: AuthConfig) -> None:
    """Clear stored tokens."""
    import contextlib

    if HAS_KEYRING:
        with contextlib.suppress(Exception):
            keyring.delete_password("rtmx", f"{config.provider}:{config.client_id}")
    token_path = _get_token_path()
    if token_path.exists():
        token_path.unlink()


# Local callback server for OAuth redirect


class _CallbackHandler(http.server.BaseHTTPRequestHandler):
    """HTTP handler for OAuth callback."""

    authorization_code: str | None = None
    error: str | None = None

    def log_message(self, format: str, *args: Any) -> None:
        """Suppress default logging."""
        pass

    def do_GET(self) -> None:
        """Handle GET request from OAuth redirect."""
        parsed = urllib.parse.urlparse(self.path)
        params = urllib.parse.parse_qs(parsed.query)

        if "code" in params:
            _CallbackHandler.authorization_code = params["code"][0]
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(b"""
<!DOCTYPE html>
<html>
<head><title>RTMX Authentication</title></head>
<body style="font-family: system-ui; text-align: center; padding: 50px;">
<h1>Authentication Successful</h1>
<p>You can close this window and return to your terminal.</p>
</body>
</html>
""")
        elif "error" in params:
            _CallbackHandler.error = params.get("error_description", params["error"])[0]
            self.send_response(400)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(
                f"""
<!DOCTYPE html>
<html>
<head><title>RTMX Authentication Error</title></head>
<body style="font-family: system-ui; text-align: center; padding: 50px;">
<h1>Authentication Failed</h1>
<p>{_CallbackHandler.error}</p>
</body>
</html>
""".encode()
            )
        else:
            self.send_response(400)
            self.end_headers()


def _wait_for_callback(port: int, timeout: int = 120) -> str | None:
    """Start callback server and wait for authorization code.

    Args:
        port: Local port to listen on
        timeout: Maximum wait time in seconds

    Returns:
        Authorization code if received, None on timeout/error
    """

    class ReuseAddrServer(socketserver.TCPServer):
        allow_reuse_address = True

    _CallbackHandler.authorization_code = None
    _CallbackHandler.error = None

    with ReuseAddrServer(("", port), _CallbackHandler) as httpd:
        httpd.timeout = timeout
        httpd.handle_request()

    if _CallbackHandler.error:
        raise AuthenticationError(_CallbackHandler.error)

    return _CallbackHandler.authorization_code


class AuthenticationError(Exception):
    """Raised when authentication fails."""

    pass


# Public API


_config: AuthConfig | None = None
_tokens: TokenInfo | None = None


def get_config() -> AuthConfig:
    """Get current authentication configuration."""
    global _config
    if _config is None:
        _config = AuthConfig()
    return _config


def set_config(config: AuthConfig) -> None:
    """Set authentication configuration."""
    global _config
    _config = config


def is_authenticated() -> bool:
    """Check if user is authenticated with valid tokens.

    Returns:
        True if authenticated with non-expired tokens
    """
    global _tokens
    config = get_config()

    if _tokens is None:
        _tokens = _load_tokens(config)

    if _tokens is None:
        return False

    # If expired but refreshable, we still consider authenticated
    return not (_tokens.is_expired and not _tokens.is_refreshable)


def get_access_token() -> str | None:
    """Get current access token, refreshing if needed.

    Returns:
        Access token string, or None if not authenticated
    """
    global _tokens
    config = get_config()

    if _tokens is None:
        _tokens = _load_tokens(config)

    if _tokens is None:
        return None

    if _tokens.is_expired and _tokens.is_refreshable:
        try:
            refresh_tokens()
        except AuthenticationError:
            return None

    return _tokens.access_token if _tokens and not _tokens.is_expired else None


async def login(open_browser: bool = True) -> TokenInfo:
    """Perform OIDC login flow.

    This is an interactive flow that:
    1. Generates PKCE codes
    2. Opens browser to authorization endpoint
    3. Waits for callback with authorization code
    4. Exchanges code for tokens
    5. Stores tokens securely

    Args:
        open_browser: Whether to automatically open browser

    Returns:
        TokenInfo with access and refresh tokens

    Raises:
        AuthenticationError: If authentication fails
    """
    global _tokens
    config = get_config()

    # Generate PKCE codes
    verifier = generate_code_verifier()
    challenge = generate_code_challenge(verifier)

    # Build authorization URL
    state = secrets.token_urlsafe(16)
    params = {
        "client_id": config.client_id,
        "response_type": "code",
        "redirect_uri": config.redirect_uri,
        "scope": " ".join(config.scopes),
        "state": state,
        "code_challenge": challenge,
        "code_challenge_method": "S256",
    }
    auth_url = f"{config.authorization_endpoint}?{urllib.parse.urlencode(params)}"

    # Start callback server in background thread
    result: dict[str, Any] = {}

    def wait_for_code() -> None:
        try:
            result["code"] = _wait_for_callback(config.callback_port)
        except AuthenticationError as e:
            result["error"] = str(e)

    thread = threading.Thread(target=wait_for_code)
    thread.start()

    # Open browser
    if open_browser:
        webbrowser.open(auth_url)
    else:
        print(f"Please open this URL in your browser:\n{auth_url}")

    # Wait for callback
    thread.join(timeout=120)

    if "error" in result:
        raise AuthenticationError(result["error"])

    if not result.get("code"):
        raise AuthenticationError("Authentication timed out or was cancelled")

    # Exchange code for tokens
    _tokens = await _exchange_code(result["code"], verifier, config)
    _store_tokens(_tokens, config)

    return _tokens


async def _exchange_code(code: str, verifier: str, config: AuthConfig) -> TokenInfo:
    """Exchange authorization code for tokens.

    Args:
        code: Authorization code from callback
        verifier: PKCE code verifier
        config: Authentication configuration

    Returns:
        TokenInfo with tokens from provider
    """
    import aiohttp

    data = {
        "grant_type": "authorization_code",
        "client_id": config.client_id,
        "code": code,
        "redirect_uri": config.redirect_uri,
        "code_verifier": verifier,
    }

    async with (
        aiohttp.ClientSession() as session,
        session.post(config.token_endpoint, data=data) as resp,
    ):
        if resp.status != 200:
            error_text = await resp.text()
            raise AuthenticationError(f"Token exchange failed: {error_text}")

        token_data = await resp.json()

    expires_in = token_data.get("expires_in", 3600)
    return TokenInfo(
        access_token=token_data["access_token"],
        refresh_token=token_data.get("refresh_token", ""),
        expires_at=datetime.now() + timedelta(seconds=expires_in),
        token_type=token_data.get("token_type", "Bearer"),
        id_token=token_data.get("id_token", ""),
    )


def refresh_tokens() -> TokenInfo:
    """Refresh access token using refresh token.

    Returns:
        Updated TokenInfo

    Raises:
        AuthenticationError: If refresh fails
    """
    import asyncio

    return asyncio.run(_refresh_tokens_async())


async def _refresh_tokens_async() -> TokenInfo:
    """Async implementation of token refresh."""
    import aiohttp

    global _tokens
    config = get_config()

    if _tokens is None or not _tokens.refresh_token:
        raise AuthenticationError("No refresh token available")

    data = {
        "grant_type": "refresh_token",
        "client_id": config.client_id,
        "refresh_token": _tokens.refresh_token,
    }

    async with (
        aiohttp.ClientSession() as session,
        session.post(config.token_endpoint, data=data) as resp,
    ):
        if resp.status != 200:
            _clear_tokens(config)
            _tokens = None
            error_text = await resp.text()
            raise AuthenticationError(f"Token refresh failed: {error_text}")

        token_data = await resp.json()

    expires_in = token_data.get("expires_in", 3600)
    _tokens = TokenInfo(
        access_token=token_data["access_token"],
        refresh_token=token_data.get("refresh_token", _tokens.refresh_token),
        expires_at=datetime.now() + timedelta(seconds=expires_in),
        token_type=token_data.get("token_type", "Bearer"),
        id_token=token_data.get("id_token", ""),
    )
    _store_tokens(_tokens, config)

    return _tokens


def logout() -> None:
    """Clear stored tokens and logout.

    This clears all stored authentication state.
    """
    global _tokens
    config = get_config()
    _clear_tokens(config)
    _tokens = None

File: internal/auth/oidc.go
// Package auth provides generic OIDC authentication with PKCE for the RTMX CLI.
//
// It works with any standards-compliant OIDC provider (Zitadel, Okta, Auth0, etc.)
// by discovering endpoints via .well-known/openid-configuration.
package auth

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// HTTPClient abstracts HTTP operations for testability.
type HTTPClient interface {
	Do(req *http.Request) (*http.Response, error)
}

// BrowserOpener abstracts opening a URL in the system browser.
type BrowserOpener func(url string) error

// OIDCClient implements the OIDC Authorization Code flow with PKCE.
type OIDCClient struct {
	Issuer       string
	ClientID     string
	Scopes       []string
	CallbackPort int

	// Injected dependencies for testability.
	HTTPClient    HTTPClient
	OpenBrowser   BrowserOpener
	TokenStorePath string

	// discoveredEndpoints caches the OIDC discovery result.
	discovery *oidcDiscovery
}

// oidcDiscovery holds endpoints from .well-known/openid-configuration.
type oidcDiscovery struct {
	AuthorizationEndpoint string `json:"authorization_endpoint"`
	TokenEndpoint         string `json:"token_endpoint"`
	UserinfoEndpoint      string `json:"userinfo_endpoint"`
	EndSessionEndpoint    string `json:"end_session_endpoint"`
	Issuer                string `json:"issuer"`
}

// TokenSet holds the tokens returned by the OIDC provider.
type TokenSet struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token,omitempty"`
	IDToken      string `json:"id_token,omitempty"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int    `json:"expires_in,omitempty"`
	ExpiresAt    int64  `json:"expires_at,omitempty"`
	Scope        string `json:"scope,omitempty"`
}

// IsExpired returns true if the access token has expired.
func (t *TokenSet) IsExpired() bool {
	if t.ExpiresAt == 0 {
		return false
	}
	return time.Now().Unix() >= t.ExpiresAt
}

// Option configures an OIDCClient.
type Option func(*OIDCClient)

// WithHTTPClient sets the HTTP client for the OIDC client.
func WithHTTPClient(c HTTPClient) Option {
	return func(o *OIDCClient) {
		o.HTTPClient = c
	}
}

// WithBrowserOpener sets the browser opener function.
func WithBrowserOpener(fn BrowserOpener) Option {
	return func(o *OIDCClient) {
		o.OpenBrowser = fn
	}
}

// WithTokenStorePath overrides the default token store path.
func WithTokenStorePath(path string) Option {
	return func(o *OIDCClient) {
		o.TokenStorePath = path
	}
}

// NewOIDCClient creates a new OIDC client with the given configuration.
func NewOIDCClient(issuer, clientID string, scopes []string, callbackPort int, opts ...Option) *OIDCClient {
	c := &OIDCClient{
		Issuer:       issuer,
		ClientID:     clientID,
		Scopes:       scopes,
		CallbackPort: callbackPort,
		HTTPClient:   http.DefaultClient,
		OpenBrowser:  defaultOpenBrowser,
	}

	for _, opt := range opts {
		opt(c)
	}

	if c.TokenStorePath == "" {
		home, _ := os.UserHomeDir()
		c.TokenStorePath = filepath.Join(home, ".rtmx", "auth", "tokens.json")
	}

	return c
}

// Discover fetches the OIDC provider configuration from the well-known endpoint.
func (c *OIDCClient) Discover(ctx context.Context) (*oidcDiscovery, error) {
	if c.discovery != nil {
		return c.discovery, nil
	}

	discoveryURL := strings.TrimRight(c.Issuer, "/") + "/.well-known/openid-configuration"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, discoveryURL, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create discovery request: %w", err)
	}

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch OIDC discovery: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("OIDC discovery returned status %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read discovery response: %w", err)
	}

	var disc oidcDiscovery
	if err := json.Unmarshal(body, &disc); err != nil {
		return nil, fmt.Errorf("failed to parse discovery response: %w", err)
	}

	if disc.AuthorizationEndpoint == "" || disc.TokenEndpoint == "" {
		return nil, fmt.Errorf("OIDC discovery missing required endpoints")
	}

	c.discovery = &disc
	return &disc, nil
}

// Login performs the full OIDC Authorization Code flow with PKCE.
//
// It opens the browser to the authorization URL, starts a local HTTP
// server to receive the callback, and exchanges the authorization code
// for tokens. The tokens are stored in the token store.
func (c *OIDCClient) Login(ctx context.Context) (*TokenSet, error) {
	disc, err := c.Discover(ctx)
	if err != nil {
		return nil, fmt.Errorf("discovery failed: %w", err)
	}

	// Generate PKCE verifier and challenge.
	verifier, err := generateCodeVerifier()
	if err != nil {
		return nil, fmt.Errorf("failed to generate code verifier: %w", err)
	}
	challenge := computeCodeChallenge(verifier)

	// Generate state and nonce for CSRF protection.
	state, err := generateRandomString(32)
	if err != nil {
		return nil, fmt.Errorf("failed to generate state: %w", err)
	}
	nonce, err := generateRandomString(32)
	if err != nil {
		return nil, fmt.Errorf("failed to generate nonce: %w", err)
	}

	// Build authorization URL.
	authURL, err := buildAuthURL(disc.AuthorizationEndpoint, c.ClientID, c.redirectURI(), c.Scopes, state, nonce, challenge)
	if err != nil {
		return nil, fmt.Errorf("failed to build auth URL: %w", err)
	}

	// Start local callback server.
	codeCh := make(chan callbackResult, 1)
	listener, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", c.CallbackPort))
	if err != nil {
		return nil, fmt.Errorf("failed to start callback server on port %d: %w", c.CallbackPort, err)
	}

	srv := &http.Server{
		Handler: callbackHandler(state, codeCh),
	}

	var wg sync.WaitGroup
	wg.Add(1)
	go func() {
		defer wg.Done()
		_ = srv.Serve(listener)
	}()

	// Open browser.
	if err := c.OpenBrowser(authURL); err != nil {
		_ = srv.Close()
		return nil, fmt.Errorf("failed to open browser: %w", err)
	}

	// Wait for callback or context cancellation.
	var result callbackResult
	select {
	case result = <-codeCh:
	case <-ctx.Done():
		_ = srv.Close()
		wg.Wait()
		return nil, ctx.Err()
	}

	// Shutdown server.
	shutdownCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	_ = srv.Shutdown(shutdownCtx)
	wg.Wait()

	if result.err != nil {
		return nil, result.err
	}

	// Exchange code for tokens.
	tokens, err := c.exchangeCode(ctx, disc.TokenEndpoint, result.code, verifier)
	if err != nil {
		return nil, fmt.Errorf("token exchange failed: %w", err)
	}

	// Set expiry timestamp if expires_in is provided.
	if tokens.ExpiresIn > 0 && tokens.ExpiresAt == 0 {
		tokens.ExpiresAt = time.Now().Unix() + int64(tokens.ExpiresIn)
	}

	// Store tokens.
	if err := c.SaveTokens(tokens); err != nil {
		return nil, fmt.Errorf("failed to save tokens: %w", err)
	}

	return tokens, nil
}

// RefreshToken uses the refresh token to get a new access token.
func (c *OIDCClient) RefreshToken(ctx context.Context) (*TokenSet, error) {
	stored, err := c.LoadTokens()
	if err != nil {
		return nil, fmt.Errorf("no stored tokens: %w", err)
	}

	if stored.RefreshToken == "" {
		return nil, fmt.Errorf("no refresh token available")
	}

	disc, err := c.Discover(ctx)
	if err != nil {
		return nil, fmt.Errorf("discovery failed: %w", err)
	}

	data := url.Values{
		"grant_type":    {"refresh_token"},
		"client_id":     {c.ClientID},
		"refresh_token": {stored.RefreshToken},
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, disc.TokenEndpoint, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create refresh request: %w", err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("refresh request failed: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read refresh response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("token refresh failed (status %d): %s", resp.StatusCode, string(body))
	}

	var tokens TokenSet
	if err := json.Unmarshal(body, &tokens); err != nil {
		return nil, fmt.Errorf("failed to parse refresh response: %w", err)
	}

	// Preserve refresh token if the provider did not return a new one.
	if tokens.RefreshToken == "" {
		tokens.RefreshToken = stored.RefreshToken
	}

	// Set expiry timestamp.
	if tokens.ExpiresIn > 0 && tokens.ExpiresAt == 0 {
		tokens.ExpiresAt = time.Now().Unix() + int64(tokens.ExpiresIn)
	}

	if err := c.SaveTokens(&tokens); err != nil {
		return nil, fmt.Errorf("failed to save refreshed tokens: %w", err)
	}

	return &tokens, nil
}

// GetAccessToken returns a valid access token, refreshing if needed.
func (c *OIDCClient) GetAccessToken(ctx context.Context) (string, error) {
	tokens, err := c.LoadTokens()
	if err != nil {
		return "", fmt.Errorf("not authenticated: %w", err)
	}

	if !tokens.IsExpired() {
		return tokens.AccessToken, nil
	}

	// Try to refresh.
	refreshed, err := c.RefreshToken(ctx)
	if err != nil {
		return "", fmt.Errorf("token expired and refresh failed: %w", err)
	}

	return refreshed.AccessToken, nil
}

// SaveTokens writes the token set to the token store file.
func (c *OIDCClient) SaveTokens(tokens *TokenSet) error {
	dir := filepath.Dir(c.TokenStorePath)
	if err := os.MkdirAll(dir, 0700); err != nil {
		return fmt.Errorf("failed to create token store directory: %w", err)
	}

	data, err := json.MarshalIndent(tokens, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal tokens: %w", err)
	}

	if err := os.WriteFile(c.TokenStorePath, data, 0600); err != nil {
		return fmt.Errorf("failed to write token store: %w", err)
	}

	return nil
}

// LoadTokens reads the token set from the token store file.
func (c *OIDCClient) LoadTokens() (*TokenSet, error) {
	data, err := os.ReadFile(c.TokenStorePath)
	if err != nil {
		return nil, fmt.Errorf("failed to read token store: %w", err)
	}

	var tokens TokenSet
	if err := json.Unmarshal(data, &tokens); err != nil {
		return nil, fmt.Errorf("failed to parse token store: %w", err)
	}

	return &tokens, nil
}

// ClearTokens removes the stored tokens.
func (c *OIDCClient) ClearTokens() error {
	err := os.Remove(c.TokenStorePath)
	if err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("failed to remove token store: %w", err)
	}
	return nil
}

// redirectURI returns the local callback URL.
func (c *OIDCClient) redirectURI() string {
	return fmt.Sprintf("http://127.0.0.1:%d/callback", c.CallbackPort)
}

// callbackResult holds the result from the OAuth callback.
type callbackResult struct {
	code string
	err  error
}

// callbackHandler returns an HTTP handler that receives the OAuth callback.
func callbackHandler(expectedState string, resultCh chan<- callbackResult) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Only handle the callback path.
		if r.URL.Path != "/callback" {
			http.NotFound(w, r)
			return
		}

		// Check for error response from provider.
		if errCode := r.URL.Query().Get("error"); errCode != "" {
			desc := r.URL.Query().Get("error_description")
			msg := fmt.Sprintf("authorization error: %s", errCode)
			if desc != "" {
				msg += ": " + desc
			}
			w.Header().Set("Content-Type", "text/plain")
			_, _ = fmt.Fprintf(w, "Authentication Failed\n\n%s\n\nYou may close this window.", msg)
			resultCh <- callbackResult{err: fmt.Errorf("%s", msg)}
			return
		}

		// Validate state.
		state := r.URL.Query().Get("state")
		if state != expectedState {
			w.Header().Set("Content-Type", "text/plain")
			_, _ = fmt.Fprint(w, "Authentication Failed\n\nState mismatch.\n\nYou may close this window.")
			resultCh <- callbackResult{err: fmt.Errorf("state mismatch: expected %q, got %q", expectedState, state)}
			return
		}

		code := r.URL.Query().Get("code")
		if code == "" {
			w.Header().Set("Content-Type", "text/plain")
			_, _ = fmt.Fprint(w, "Authentication Failed\n\nNo authorization code received.\n\nYou may close this window.")
			resultCh <- callbackResult{err: fmt.Errorf("no authorization code in callback")}
			return
		}

		w.Header().Set("Content-Type", "text/plain")
		_, _ = fmt.Fprint(w, "Authentication Successful\n\nYou may close this window and return to the CLI.")
		resultCh <- callbackResult{code: code}
	})
}

// exchangeCode exchanges an authorization code for tokens.
func (c *OIDCClient) exchangeCode(ctx context.Context, tokenEndpoint, code, verifier string) (*TokenSet, error) {
	data := url.Values{
		"grant_type":    {"authorization_code"},
		"client_id":     {c.ClientID},
		"code":          {code},
		"redirect_uri":  {c.redirectURI()},
		"code_verifier": {verifier},
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, tokenEndpoint, strings.NewReader(data.Encode()))
	if err != nil {
		return nil, fmt.Errorf("failed to create token request: %w", err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := c.HTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("token request failed: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read token response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("token exchange failed (status %d): %s", resp.StatusCode, string(body))
	}

	var tokens TokenSet
	if err := json.Unmarshal(body, &tokens); err != nil {
		return nil, fmt.Errorf("failed to parse token response: %w", err)
	}

	return &tokens, nil
}

// buildAuthURL constructs the OIDC authorization URL.
func buildAuthURL(endpoint, clientID, redirectURI string, scopes []string, state, nonce, codeChallenge string) (string, error) {
	u, err := url.Parse(endpoint)
	if err != nil {
		return "", fmt.Errorf("invalid authorization endpoint: %w", err)
	}

	q := u.Query()
	q.Set("response_type", "code")
	q.Set("client_id", clientID)
	q.Set("redirect_uri", redirectURI)
	q.Set("scope", strings.Join(scopes, " "))
	q.Set("state", state)
	q.Set("nonce", nonce)
	q.Set("code_challenge", codeChallenge)
	q.Set("code_challenge_method", "S256")
	u.RawQuery = q.Encode()

	return u.String(), nil
}

// generateCodeVerifier creates a cryptographically random PKCE code verifier.
// Per RFC 7636, the verifier is 43-128 characters from [A-Z][a-z][0-9]-._~
func generateCodeVerifier() (string, error) {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

// computeCodeChallenge computes the S256 code challenge from a verifier.
func computeCodeChallenge(verifier string) string {
	h := sha256.Sum256([]byte(verifier))
	return base64.RawURLEncoding.EncodeToString(h[:])
}

// generateRandomString creates a URL-safe random string of the given byte length.
func generateRandomString(n int) (string, error) {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

// defaultOpenBrowser is a no-op in production; overridden for real usage.
// Real browser opening is injected via WithBrowserOpener.
func defaultOpenBrowser(url string) error {
	return fmt.Errorf("browser opener not configured; visit: %s", url)
}
